lib/realtime/tenants/authorization/permissions/broadcast_permissions.ex

```diff
@@ -0,0 +1,93 @@
+  alias Realtime.Tenants.Authorization.Permissions
+
+  defstruct read: false, write: false
+
+  @behaviour Realtime.Tenants.Authorization.Permissions
+
+  @type t :: %__MODULE__{
+          :read => boolean(),
+          :write => boolean()
+        }
+  @impl true
+  def check_read_permissions(_conn, permissions, %Authorization{channel: nil}) do
+    {:ok, Permissions.update_permissions(permissions, :broadcast, :read, false)}
+  end
+
+  def check_read_permissions(conn, %Permissions{} = permissions, %Authorization{
+        channel: %Channel{id: channel_id}
+      }) do
+    query = from(b in Broadcast, where: b.channel_id == ^channel_id, select: b.channel_id)
+
+    case Repo.all(conn, query, Broadcast, mode: :savepoint) do
+      {:ok, broadcast} when broadcast != [] ->
+        permissions = Permissions.update_permissions(permissions, :broadcast, :read, true)
+        {:ok, permissions}
+
+      {:ok, _} ->
+        permissions = Permissions.update_permissions(permissions, :broadcast, :read, false)
+        {:ok, permissions}
+
+      {:error, %Postgrex.Error{postgres: %{code: :insufficient_privilege}}} ->
+        permissions = Permissions.update_permissions(permissions, :broadcast, :read, false)
+        {:ok, permissions}
+
+      {:error, error} ->
+        Logger.error("Error getting permissions for connection: #{inspect(error)}")
+        {:error, error}
+    end
+  end
+
+  @impl true
+  def check_write_permissions(_conn, permissions, %Authorization{channel: nil}) do
+    {:ok, Permissions.update_permissions(permissions, :broadcast, :write, false)}
+  end
+
+  def check_write_permissions(conn, permissions, %Authorization{
+        channel: %Channel{id: channel_id}
+      }) do
+    query = from(b in Broadcast, where: b.channel_id == ^channel_id)
+
+    with {:ok, broadcast} <- Repo.one(conn, query, Broadcast),
+         changeset <- Broadcast.check_changeset(broadcast, %{check: true}),
+         {:ok, %Broadcast{check: true} = broadcast} <-
+           Repo.update(conn, changeset, Broadcast, mode: :savepoint) do
+      revert_changeset = Broadcast.check_changeset(broadcast, %{check: false})
+      {:ok, _} = Repo.update(conn, revert_changeset, Broadcast)
+      permissions = Permissions.update_permissions(permissions, :broadcast, :write, true)
+
+      {:ok, permissions}
+    else
+      {:ok, _} ->
+        permissions = Permissions.update_permissions(permissions, :broadcast, :write, false)
+        {:ok, permissions}
+
+      {:error, %Postgrex.Error{postgres: %{code: :insufficient_privilege}}} ->
+        permissions = Permissions.update_permissions(permissions, :broadcast, :write, false)
+        {:ok, permissions}
+
+      {:error, :not_found} ->
+        permissions = Permissions.update_permissions(permissions, :broadcast, :write, false)
+
+        {:ok, permissions}
+
+      {:error, error} ->
+        Logger.error("Error getting permissions for connection: #{inspect(error)}")
+        {:error, error}
+    end
```

Comment:
This with/1 has 2 "<-" clauses that can "fail" and 4 "else" clauses that handle errors from the two. In this particular case I think with/1 is more of a hindrance than help because one needs to jump between when the error may originate and when it is handled. I'd give a nested case a shot and see if it's readable enough. WDYT?

Reply:
Good call. I do think that it looks odd and hurts visibility and might even find better ways to abstract some of this stuff since it feels something here might "repeat itself"..

Reply:
looks way better in terms of readability and will for sure be more maintainable and less "magic"